### PR TITLE
Add SolverGCR

### PR DESCRIPTION
**Pull request by peterrum**
I am working on this together with @dravinsi.

### CODE REVIEW

**Comment by tjhei** [comment]
Do you have a good reference for the algorithm?

File: include/deal.II/lac/solver_gcr.h [hunk 1]
```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <deal.II/lac/solver.h>
+
+namespace dealii
```
Comment (bangerth) [inline include/deal.II/lac/solver_gcr.h:5]: There are a lot of issues with these first five lines of the file :-)

Comment (peterrum) [reply include/deal.II/lac/solver_gcr.h:5]: There is a reason why the PR is WIP ;)

**Comment by dravinsi** [comment]
> Do you have a good reference for the algorithm?

I apologize for the late reply. The original GCR is mentioned the GMRES paper by Saad and Schultz, found here https://web.stanford.edu/class/cme324/saad-schultz.pdf 

A preconditioned variant of GCR (referred to as GMRESR)  is described in a paper by H. A. Van der Vorst & C. Vuik, found here https://onlinelibrary.wiley.com/doi/abs/10.1002/nla.1680010404

The latter reference was obtained from Y. Notay's paper on AMG found here https://etna.math.kent.edu/vol.37.2010/pp123-146.dir/pp123-146.pdf where it is reference [40]. This paper also includes a compact pseudocode. 

I will also stress what Peter has said; that this is a work in progress.

File: include/deal.II/lac/solver_gcr.h [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  A.vmult(p, x);
+  p.add(-1., b);
+
+  if (use_default_residual == true)
+    res = p.l2_norm();
+
+  preconditioner.vmult(search, p);
+
+  if (use_default_residual == false)
+    res = search.l2_norm();
```
Comment (peterrum) [inline include/deal.II/lac/solver_gcr.h:147]: residuum vs. preconditioned residuum

File: include/deal.II/lac/solver_gcr.h [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+  conv = this->iteration_status(it, res, x);
+  if (conv != SolverControl::iterate)
+    return;
+
+  while (conv == SolverControl::iterate)
+    {
+      it++;
+
+      const unsigned int it_ = ((it - 1) % max_n_tmp_vectors) + 1; // restart
```
Comment (peterrum) [inline include/deal.II/lac/solver_gcr.h:159]: an attempt for restart if the number of iterations is higher than max_n_tmp_vectors

**Review by peterrum — COMMENTED** [review]
@dravinsi Could you take a look at the code and try it out? Thx!

**Pull request closed without merging**